tests/test_modeling_common.py

```diff
@@ -1662,6 +1664,34 @@ def test_can_use_safetensors(self):
                         f"The shared pointers are incorrect, found different pointers for keys {shared_names}",
                     )
 
+    def test_tied_weights_keys(self):
+        config, _ = self.model_tester.prepare_config_and_inputs_for_common()
+        config.tie_word_embeddings = True
+        for model_class in self.all_model_classes:
+            model_tied = model_class(config)
+
+            ptrs = collections.defaultdict(list)
+            for name, tensor in model_tied.state_dict().items():
+                ident = (tensor.data_ptr(), tensor.device, tensor.shape, tensor.stride())
```

Comment:
This will only catch  exactly shared tensors, never partial of transposed sharing.

We shouldn't have it in our code, but since this is a test, maybe making it more robust could be good (so we can catch whenever a new model with finer sharing comes along).

`(tensor.untyped_storage().data_ptr(), tensor.untyped_storage().nbytes())` is more robust. (There's no stride, nor shape on storage). (`untyped_storage` is 2.0 otherwise you have to use `.storage().bytes()` iirc.

This doesn't actually capture shared/non-shared storages faking (@thomasw21) but I think it's beyond the point.
There are ways to make 2 different storages actually point to the same memory (without necessary sharing either pointer nor size), leading to all sorts of weird things.

Reply:
I guess this is fine here since the `_tied_weights_keys` mechanism shares not only the storage but also the entire `nn.Parameter` no?

Reply:
Also I guess you could just use `id` no?

Reply:
>  So maybe let's factor that in a util function we reuse for this test, the safetensors check for duplicates and the checkpoint sharding if that's the strongest way?

Seems good to me. At least we have a unique source for pointer sharing.

Reply:
> The simple id check is not sufficient @thomasw21 : it doesn't catch the tied weights that DeformableDetr has (I compared the output for all models).

Hum that seems very suspicious to me. I've had issues before where if you had 2 parameters (with different ids) that share the underlying tensors/storage ... then gradients where pretty weird ...

Reply:
The corresponding model is used so little it wouldn't surprise me if training it was not working. That's [how](https://github.com/huggingface/transformers/blob/3723329d014a7b144863e597ea4fe6de5e6a8279/src/transformers/models/deformable_detr/modeling_deformable_detr.py#L308) it clones the weights, which is indeed pretty ugly

Reply:
Ah very good find!
```python
import copy

from torch import nn

def main():
    model = nn.Linear(3,5)
    new_model = copy.deepcopy(model)
    # Theses are different parameters ....
    assert id(model.weight) != id(new_model.weight)
    
if __name__ == "__main__":
    main()
```